File: solutions/fsharp/queen-attack/1/QueenAttack.fs
module QueenAttack

open System

let (|OnBoard|_|) (position: int * int) =
   let x = fst position
   let y = snd position
   let onBoard n = n > -1 && n < 8
   if (onBoard x && onBoard y) then Some(position) else None

let create (position: int * int) =
   match position with
   | OnBoard _ -> true
   | _         -> false

let distance (x:int) (y:int) = x - y |> Math.Abs

let canAttack (queen1: int * int) (queen2: int * int) =
   let row     = fst queen1 = fst queen2
   let column  = snd queen1 = snd queen2
   let diagonal = 
      let distanceX = distance (fst queen1) (fst queen2)
      let distanceY = distance (snd queen1) (snd queen2)
      distanceX = distanceY

   row || column || diagonal

File: solutions/csharp/queen-attack/1/QueenAttack.cs
public class Queen
{
    public Queen(int row, int column)
    {
        if (row < 0 || column < 0 || row > 7 || column > 7)
            throw new ArgumentOutOfRangeException();

        Row = row;
        Column = column;
    }

    public int Row { get; }
    public int Column { get; }
}

public static class QueenAttack
{
    static int Distance(int lhs, int rhs) => Math.Abs(lhs - rhs);

    public static Queen Create(int row, int column) => new(row, column);

    public static bool CanAttack(Queen white, Queen black)
    {
        if (white.Column == black.Column) return true;
        if (white.Row == black.Row) return true;
        if (Distance(white.Row, black.Row)
            == Distance(white.Column, black.Column)) return true;
        return false;
    }
}